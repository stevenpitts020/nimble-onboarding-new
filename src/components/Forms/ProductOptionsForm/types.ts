import React from "react";
import { IProductOption as IProductOptionProspect } from "../../../store/reducers/type";

export interface IProductOptionItem {
  key: string;
  annotation?: string;
  title: string;
  lead?: string;
  value: string;
}
export interface IProductOption {
  className?: string;
  style?: React.CSSProperties;
  option: IProductOptionItem;
  handleChange: (ev: React.ChangeEvent<HTMLInputElement>) => void;
  isChildren?: boolean;
}
export interface IProductOptionsForm {
  className?: string;
  style?: React.CSSProperties;
  options: IProductOptionProspect[];
  onSubmit: (data: IProductOptionProspect[]) => void;
  onBack: () => void;
}
