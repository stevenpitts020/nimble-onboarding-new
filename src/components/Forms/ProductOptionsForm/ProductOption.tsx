import React from "react";
import { IProductOption } from "./types";

const ProductOption: React.FC<IProductOption> = (props) => {
  const option = props.option;
  return (
    <div
      key={option.key}
      className={`product-option-item ${props.className} ${
        props.isChildren ? "sub-option" : ""
      } ${
        option.annotation === "Can be ordered thru online banking"
          ? "no-option"
          : ""
      }`}
    >
      <div className="option_desc">
        <p>{option.title}</p>
        <span>{option.lead}</span>
      </div>
      {option.annotation !== "Can be ordered thru online banking" ? (
        <div className="switch-wrapper">
          <p className={`price ${option.annotation === "Free" ? "free" : ""}`}>
            {option.annotation}
          </p>
          <label className="switch">
            <input
              type="checkbox"
              name={option.key}
              onChange={props.handleChange}
              data-testid={`checkbox-${option.key}`}
              checked={option.value === "true"}
            />
            <span className="slider round" />
          </label>
        </div>
      ) : (
        <span className="annotation">{option.annotation}</span>
      )}
    </div>
  );
};
export default ProductOption;
