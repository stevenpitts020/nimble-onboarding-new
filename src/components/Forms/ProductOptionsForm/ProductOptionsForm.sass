@import '../../../styles/common'

$colorSwitch: #5CB97F
$colorPrice: #F19920

form.product-options-wrapper
  display: flex
  flex-flow: column
  width: 100%
  margin-top: 48px
  .option-wrapper
    &.last
      order: 1
    .product-option-item
      display: flex
      justify-content: space-between
      align-items: center
      border-bottom: solid 1px $superLightGray
      padding: 32px 0
      position: relative
      order: 0
      &.last
        order: 1
      &.sub-option
        .option_desc
          margin-left: 40px
          max-width: 270px
      &.no-option
      .option_desc
        text-align: left
        position: relative
        p
          font-weight: 700
          font-size: 20px
          line-height: 24px
          margin: 0
      span
        font-size: 14px
        color: $text-color
        &.annotation
          right: 0
      .price
        margin: 0
        font-size: 14px
        font-weight: 600
        color: $colorPrice
        margin-right: 60px
        &.free
          color: $colorSwitch
      .switch-wrapper
        display: flex
        .switch
          display: flex
          margin: 0
  .form-controls
    margin-top: 48px
    order: 3
    display: flex
    justify-content: space-between

  /* The switch box around the slider */
  .switch
    position: relative
    display: inline-block
    min-width: 48px
    width: 48px
    height: 24px

  /* Hide checkbox */
  .switch input
    opacity: 0
    width: 0
    height: 0

  /* slider */
  .slider
    position: absolute
    cursor: pointer
    top: 0
    left: 0
    right: 0
    bottom: 0
    background-color: #ccc
    -webkit-transition: .4s
    transition: .4s

  .slider:before
    position: absolute
    content: ""
    height: 16px
    width: 16px
    left: 6px
    bottom: 4px
    background-color: #ffffff
    -webkit-transition: .4s
    transition: .4s

  input:checked + .slider
    background-color: $colorSwitch

  input:focus + .slider
    box-shadow: 0 0 1px $colorSwitch

  input:checked + .slider:before
    transform: translateX(22px)

  /* Rounded sliders */
  .slider.round
    border-radius: 34px
    &:before
      border-radius: 50%
