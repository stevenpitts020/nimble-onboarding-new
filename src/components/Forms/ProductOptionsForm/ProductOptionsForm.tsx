import React, { useState } from "react";
import { ArrowLeft } from "react-feather";
import { useForm } from "react-hook-form";
import "./ProductOptionsForm.sass";
import ProductOption from "./ProductOption";
import { log } from "../../../services";
import { IProductOptionsForm } from "./types";
import { IProductOption } from "../../../store/reducers/type";
import Button from "../../Common/Button/Button";

const ProductOptionsForm: React.FC<IProductOptionsForm> = (props) => {
  const { handleSubmit, formState } = useForm<IProductOption>({
    mode: "onChange",
  });
  const { onSubmit } = props;
  const optionsInStr = sessionStorage.getItem("PRODUCT_OPTIONS");
  const options = optionsInStr ? JSON.parse(optionsInStr) : props.options;
  const prepareOptions = () =>
    options.map((option: IProductOption) => ({
      id: option.id,
      key: option.key,
      value: option.value !== null ? option.value : "null",
      category: option.category,
      title: option.title,
      children: props.options.filter((children: IProductOption) => {
        if (
          children.parentId !== undefined &&
          children.parentId === option.id
        ) {
          return { ...children, parentId: children.parentId, value: "true" };
        }
        return null;
      }),
      isChildren: !!option.parentId,
      parentId: option.parentId,
      lead: option.lead ? option.lead : "",
      annotation: option.annotation ? option.annotation : "",
    }));
  const [formProducts, setFormProducts] = useState(prepareOptions());
  const onFormSubmit = async () => {
    const dataToSubmit = formProducts
      .map((option: IProductOption) => ({
        key: option.key,
        value: option.value,
        category: option.category,
        title: option.title,
      }))
      .filter((option: IProductOption) => option.value !== "null");
    sessionStorage.setItem("PRODUCT_OPTIONS", JSON.stringify(formProducts));
    log.info(dataToSubmit, "update state with product options");
    onSubmit(dataToSubmit);
  };

  const goBack = () => {
    props.onBack();
  };

  const handleChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
    log.info("HITG", "handleChange");
    const propertyKey: string = ev.target.name;
    const newState = [...formProducts];
    const i = newState.findIndex((e) => e.key === propertyKey);

    if (i !== -1) {
      newState[i].value = newState[i].value === "true" ? "false" : "true";

      log.info(newState[i], "handleChangeTG");
      log.info(i, "i");

      if (newState[i].children.length >= 1) {
        // if parent option is false so should the children options be
        log.info("update children options", "handleChange");

        newState[i].children.forEach((children: any) => {
          const childrenIndex = newState.findIndex(
            (el) => el.key === children.key
          );
          // if we deactivate parent, we deactivate children
          // also: there is a second object so I'm syncing that as well
          newState[childrenIndex].value = newState[i].value;
          children.value = newState[i].value;
        });
      }

      // handle children option change
      if (newState[i].isChildren) {
        const parentIndex = newState.findIndex(
          (el) => el.id === newState[i].parentId
        );
        log.info("is children", "handleChange");

        newState[parentIndex].children.forEach((children: any) => {
          if (children.key === propertyKey) {
            log.info("sync parent option", "handleChange");
            children.value = newState[i].value;
          }
        });
      }

      // relationships
      switch (propertyKey) {
        // e_statments should be 'true' if mobile_and_online_banking is selected
        case "mobile_and_online_banking":
          if (newState[i].value === "true") {
            newState[
              newState.findIndex((e) => e.key === "e_statements")
            ].value = "true";
          } else {
            newState[
              newState.findIndex((e) => e.key === "e_statements")
            ].value = "false";
          }
          break;
        // paper_statements should be 'false' if e_statments is selected
        case "e_statements":
          if (newState[i].value === "true") {
            newState[
              newState.findIndex((e) => e.key === "paper_statements")
            ].value = "false";
          }
          if (newState[i].value === "false") {
            newState[
              newState.findIndex((e) => e.key === "paper_statements")
            ].value = "true";
          }
          break;
        case "paper_statements":
          if (newState[i].value === "true") {
            newState[
              newState.findIndex((e) => e.key === "e_statements")
            ].value = "false";
          }
          if (newState[i].value === "false") {
            newState[
              newState.findIndex((e) => e.key === "e_statements")
            ].value = "true";
          }
          break;
      }

      log.info(newState, "handleChange");

      setFormProducts(newState);
    }
  };

  return (
    <div className="ni-test prospect-form-card" data-testid="BSAForm">
      <form
        onSubmit={handleSubmit(onFormSubmit)}
        className="product-options-wrapper"
      >
        {formProducts.length >= 1 ? (
          formProducts.map((option: any) => {
            if (option.isChildren === false) {
              return (
                <div
                  key={option.key}
                  className={`option-wrapper ${
                    option.key === "checks" ? "last" : ""
                  }`}
                >
                  <ProductOption
                    option={option}
                    handleChange={handleChange}
                    key={option.key}
                  />
                  {option.children?.map((suboption: any) => {
                    if (option.value === "true") {
                      return (
                        <ProductOption
                          option={suboption}
                          handleChange={handleChange}
                          isChildren
                          key={suboption.key}
                        />
                      );
                    }
                  })}
                </div>
              );
            }
            return null;
          })
        ) : (
          <h3>No additional features available</h3>
        )}
        <div className="form-controls">
          <Button className="back" onClick={goBack} data-testid="back-btn">
            <ArrowLeft color="#444647" />
            Back
          </Button>
          <Button
            disabled={!formState.isValid}
            type="submit"
            className="is-pill is-green"
            data-testid="submit-btn"
          >
            Proceed
          </Button>
        </div>
      </form>
    </div>
  );
};
export default ProductOptionsForm;
