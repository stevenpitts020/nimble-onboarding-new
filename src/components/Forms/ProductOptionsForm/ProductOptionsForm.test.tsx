import React from "react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { render, fireEvent, act, screen } from "@testing-library/react";
import ProductOptionsForm from "./ProductOptionsForm";

describe("ProductOptionsForm", () => {
  const options = [
    {
      id: "99416984-51dd-47e2-8dca-470872a6ba96",
      key: "atm_only_card",
      category: "product_features",
      title: "ATM Only Card",
      value: "false",
      annotation: "Free",
    },
    {
      id: "b428e1c8-cbcd-4e8a-b164-975ca7b9b4a4",
      key: "bill_pay",
      category: "product_features",
      title: "Bill Pay",
      value: "false",
      annotation: "Free",
      parentId: "e8823979-9550-416c-a5fb-f9d6aad3fcde",
    },
    {
      id: "75b0c309-1412-4786-ae42-8fbbfa5dc34c",
      key: "checks",
      category: "product_features",
      title: "Checks",
      value: "false",
      annotation: "Can be ordered thru online banking",
    },
    {
      id: "c510879c-da2d-4f25-96cf-028f6ada669b",
      key: "debit_card",
      category: "product_features",
      title: "Debit Card",
      value: "false",
      lead: "ShazamÂ®Chek Card",
      annotation: "Free",
    },
    {
      id: "ddc32aab-4d05-4c20-a1d6-6d5034433c2a",
      key: "e_statements",
      category: "product_features",
      title: "E-Statements",
      value: "false",
      lead: "Get your bank statements directly on your email",
    },
    {
      id: "e8823979-9550-416c-a5fb-f9d6aad3fcde",
      key: "mobile_and_online_banking",
      category: "product_features",
      title: "Mobile & Online Banking",
      value: "false",
      annotation: "Free",
    },
    {
      id: "5ac2e4d0-a0a2-43d6-bcc8-00b7e57a3b13",
      key: "mobile_deposits",
      category: "product_features",
      title: "Mobile Deposits",
      value: "false",
      annotation: "Free",
      parentId: "e8823979-9550-416c-a5fb-f9d6aad3fcde",
    },
  ];
  const history = createMemoryHistory();
  const onSubmit = jest.fn();
  const onBack = jest.fn();

  test("should show all the product options", () => {
    render(
      <Router history={history}>
        <ProductOptionsForm
          onSubmit={onSubmit}
          options={options}
          onBack={onBack}
        />
      </Router>
    );
    expect(screen.getAllByRole("checkbox")).toMatchSnapshot();
  });
  test("clicking Back Button should call onBack", async () => {
    const { getByTestId } = render(
      <Router history={history}>
        <ProductOptionsForm
          onSubmit={onSubmit}
          options={options}
          onBack={onBack}
        />
      </Router>
    );
    await act(async () => {
      fireEvent.click(getByTestId("back-btn"));
    });
    expect(onBack).toBeCalled();
  });
  test("after submitting it call onSubmit", async () => {
    const { getByTestId } = render(
      <Router history={history}>
        <ProductOptionsForm
          onSubmit={onSubmit}
          options={options}
          onBack={onBack}
        />
      </Router>
    );
    await act(async () => {
      fireEvent.click(getByTestId("submit-btn"));
    });
    expect(onSubmit).toHaveBeenCalled();
  });
});
