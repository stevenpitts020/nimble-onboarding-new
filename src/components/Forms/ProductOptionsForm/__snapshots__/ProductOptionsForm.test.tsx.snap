// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductOptionsForm should show all the product options 1`] = `
Array [
  <input
    data-testid="checkbox-atm_only_card"
    name="atm_only_card"
    type="checkbox"
  />,
  <input
    data-testid="checkbox-debit_card"
    name="debit_card"
    type="checkbox"
  />,
  <input
    data-testid="checkbox-e_statements"
    name="e_statements"
    type="checkbox"
  />,
  <input
    data-testid="checkbox-mobile_and_online_banking"
    name="mobile_and_online_banking"
    type="checkbox"
  />,
]
`;
